src/main/java/ir/ac/sbu/controller/MainController.java

```diff
@@ -419,4 +440,27 @@ private void showModal(URL resource, String title) {
             e.printStackTrace();
         }
     }
+
+    private void setOnCloseRequest() {
+        Main.getPrimaryStage().setOnCloseRequest(event -> {
+            windowEvent = event;
+            String dialogString = lastChosenFileForSave != null ? "Do you want to save changes?" : "Do you want to save changes to a file?";
+            Optional<ButtonType> result = DialogUtility.showConfirmationDialog("PGEN", dialogString);
+            if (!result.isPresent()) {
+                return;
```

Comment:
It means window closed. Do this here:
```windowEvent.consume()```
By doing this you do not need `windowEvent` variable in class level. remove it.